include/device_queue.h

```diff
@@ -59,9 +64,166 @@ namespace detail {
 		std::unique_ptr<cl::sycl::queue> sycl_queue;
 		bool device_profiling_enabled = false;
 
-		cl::sycl::device pick_device(const config& cfg, cl::sycl::device* user_device) const;
 		void handle_async_exceptions(cl::sycl::exception_list el) const;
 	};
 
+	// Try to find a platform that can provide a unique device for each node using a device selector.
+	template <typename DeviceT, typename PlatformT, typename SelectorT>
+	bool try_find_device_per_node(std::string& how_selected, DeviceT& device, const std::vector<PlatformT>& platforms, const host_config& host_cfg,
+	    SelectorT selector, cl::sycl::info::device_type type) {
+		std::vector<std::tuple<DeviceT, size_t>> devices_with_platform_idx;
+		for(size_t i = 0; i < platforms.size(); ++i) {
+			auto&& platform = platforms[i];
+			if(platform.get_devices().size() >= host_cfg.node_count) {
+				for(auto device : platform.get_devices()) {
+					if(selector(device) == -1) { continue; }
+					devices_with_platform_idx.emplace_back(device, i);
+				}
+			}
+		}
+
+		std::sort(devices_with_platform_idx.begin(), devices_with_platform_idx.end(),
+		    [selector](const auto& a, const auto& b) { return selector(std::get<0>(a)) > selector(std::get<0>(b)); });
+		bool same_platform = true;
+		if(devices_with_platform_idx.size() >= host_cfg.node_count) {
+			auto [device_from_platform, idx] = devices_with_platform_idx[0];
+			auto platform_name = device_from_platform.get_platform().template get_info<cl::sycl::info::platform::name>();
+
+			for(size_t i = 1; i < host_cfg.node_count; ++i) {
+				auto [device_from_platform, idx] = devices_with_platform_idx[i];
+				if(device_from_platform.get_platform().template get_info<cl::sycl::info::platform::name>() != platform_name) { same_platform = false; }
```

Comment:
So really the behavior here is that a selection is only made if the first N devices, according to their selector score, are from the same platform. That is fine in principle I guess, however should be tested explicitly. Also, devices that have the same score but belong to different platforms could cause problems here (while being difficult to test, depending on stability of `std::sort`).

Reply:
From my understanding, the test "pick_device doesn't choose devices across platform" tests the same thing. Also, we could use the `std::stable_sort` here, however, it is not memory efficient.

Reply:
Okay yes, that test does indeed assert that behavior. However I'm wondering whether this is actually what we want. Because if we for example wanted to select 2 devices and had a score/platform array of the form `[(1, 0), (1, 1), (1, 1)]`, because platform `0` happened to only have one device with a non-negative score, we end up selecting nothing, even though platform `1` could supply two devices with the same score (`1`). While I don't think we should overcomplicate things, in this case the outcome of the selection only depends on the implementation-defined order in which platforms are enumerated, which also doesn't seem right.

And yes, I think we should be using `std::stable_sort` here!

Reply:
Discussion today:
Let's select the top N devices, regardless of platform.
Warn the user iff (i) the devices are from different platforms or (ii) they are different device types.